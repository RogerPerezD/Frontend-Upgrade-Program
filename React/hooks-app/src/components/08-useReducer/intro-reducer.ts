import React from 'react';


type State = {
    id: number,
    todo: string,
    done: boolean
};

type Action = {
    type: string,
    payload: any,
};

const initialState: State[] = [{
    id: 1,
    todo: 'Buy bread',
    done: false
}];


const todoReducer = ( state: State [] = initialState, action?: Action ) =>{
    if (action?.type === 'add') {
        return [ ...state, action.payload];
    }

    return state;
}

let todos = todoReducer();

console.log(todos);

const newTodo: State = {
    id: 2,
    todo: 'Buy milk',
    done: false
}

const todoAddAction: Action = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer( todos, todoAddAction);

console.log(todos);



