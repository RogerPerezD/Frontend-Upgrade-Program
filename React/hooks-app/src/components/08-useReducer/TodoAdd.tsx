import React, { FormEvent, useState } from 'react';
import { Todo } from './TodoApp';

type TodoAddProps= {
    handleAddTodo( newTodo: Todo): void
}

export const TodoAdd = ({ handleAddTodo }:TodoAddProps) => {
    const [inputState, setInputState] = useState<string>('');

    const handleInput = (e: FormEvent<HTMLInputElement>) =>{
        setInputState( e.currentTarget.value );
    }

    const handleSubmit = (e: FormEvent<HTMLFormElement>) =>{
        e.preventDefault();

        // Validar que no este vacia la caja
        if (inputState.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: inputState,
            done: false
        }

        handleAddTodo( newTodo );

        setInputState('');
    }

    return (
        <form onSubmit={ handleSubmit }>
            <input type="text" 
                className='form-control' 
                name='description' 
                placeholder='Learn....' 
                autoComplete='off'
                onChange={ handleInput }
                value={ inputState }/>
            <div className="d-grid gap-1">
            <button 
                className='btn btn-outline-primary mt-1' 
                type='submit'> 
                Add 
            </button>
            </div>
        </form>
    );
};
