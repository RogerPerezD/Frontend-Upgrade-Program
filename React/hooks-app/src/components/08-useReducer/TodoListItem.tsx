import React, { MouseEvent } from 'react';
import { Todo } from './TodoApp';

type TodoListItemProps = {
    todo: Todo,
    index: number,
    handleToggleTodo(e: MouseEvent<HTMLParagraphElement>, todo: Todo): void,
    handleDeleteTodo( todo: Todo): void
}

export const TodoListItem = React.memo(({ todo, index, handleToggleTodo, handleDeleteTodo }: TodoListItemProps) => {
    console.log('list item renderizado');
    return (
    <li className='list-group-item'>
        <p 
        className='text-center'
        onClick={ (e) => { handleToggleTodo( e, todo )} }>{index}. { todo.desc }</p>
        
        <button 
        className='btn btn-danger'
        onClick={ () => { handleDeleteTodo (todo)} }>
            Delete
        </button>
    </li>
  );
});
