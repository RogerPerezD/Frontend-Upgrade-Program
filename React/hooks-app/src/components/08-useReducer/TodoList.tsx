import React, { MouseEvent } from 'react';
import { Todo } from './TodoApp';
import { TodoListItem } from './TodoListItem';

type TodoListProps = {
    todos: Todo[],
    handleToggleTodo(e: MouseEvent<HTMLParagraphElement>, todo: Todo): void,
    handleDeleteTodo( todo: Todo): void
}

export const TodoList = React.memo(({ todos, handleToggleTodo, handleDeleteTodo}: TodoListProps) => {
    console.log('Todo list renderizado');
    return (
        <ul className='list-group list-group-flush'>
            {
                todos.map( (todo, i) => ( 
                <TodoListItem 
                key={todo.id}
                todo={ todo }
                index={ i+1 }
                handleToggleTodo = { handleToggleTodo }
                handleDeleteTodo = { handleDeleteTodo }
                />
                ))
            }
            
        </ul>  
    );
});
