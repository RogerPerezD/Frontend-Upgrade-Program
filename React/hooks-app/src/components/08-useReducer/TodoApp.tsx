import React, { MouseEvent, useCallback, useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';

import './style.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

export type Todo = {
    id: number,
    desc: string,
    done: boolean
}

const init = () =>{
    return JSON.parse(localStorage.getItem('todos')!) || [];
}

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [],init);
    
    useEffect(() => {
        // Convertimos los todos en un Json en texto plano
        // Ya que el local storage solo puede guardar strings
        localStorage.setItem( 'todos', JSON.stringify( todos ));
    }, [todos]);
    
    const handleAddTodo = ( newTodo: Todo) =>{
        const action = {
            type: 'add',
            payload: newTodo,
        }
        dispatch( action );
    }

    const handleDeleteTodo = useCallback((todo: Todo) => {
        const action = {
            type: 'delete',
            payload: todo
        }
        dispatch( action );
    }, []);

    const handleToggleTodo = useCallback((e: MouseEvent<HTMLParagraphElement>, todo: Todo) => {
        const action = {
            type: 'update',
            payload: todo
        }
        dispatch( action );
        e.currentTarget.classList.toggle('complete');
    }, []);


    return (
        <div>
            <h1>TodoApp ({ todos.length })</h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <TodoList 
                    todos={ todos } 
                    handleToggleTodo={ handleToggleTodo }
                    handleDeleteTodo = { handleDeleteTodo }
                    />
                </div>

                <div className="col-5">
                    <h4>Add Todo</h4>
                    <hr />
                    <TodoAdd handleAddTodo={ handleAddTodo }/>
                </div>
            </div>
        </div>
    );
};
