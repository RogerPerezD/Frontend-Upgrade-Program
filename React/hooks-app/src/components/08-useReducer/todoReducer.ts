import { Todo } from "./TodoApp";

type Action = {
    type: string,
    payload: Todo,
};

// Inicializamos un arreglo vacio del tipo Todo
export const todoReducer = ( state = ([] as Todo []), action: Action): Todo[] => {
  
    switch (action.type) {
        case 'add':
            return [ ...state, action.payload];
        case 'delete':
            return state.filter( (todo) => todo.id !== action.payload.id);
        case 'update':
            return state.map( todo => ( 
                todo.id === action.payload.id 
                ? {...todo, done: !todo.done} 
                : todo
                ));
        // case 'update-old':
        // return state.map( (todo) => {
        //     if (todo.id === action.payload.id) {
        //         return {...todo, done: !todo.done}
        //     }
        //     return todo;
        // });
        default:
            return state;
    }
};
