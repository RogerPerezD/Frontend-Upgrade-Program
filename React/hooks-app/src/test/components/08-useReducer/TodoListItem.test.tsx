import { shallow } from "enzyme";
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';


describe('Test in <TodoListItem/>', () => {
    const handleToggleTodo = jest.fn();
    const handleDeleteTodo = jest.fn();
    const wrapper = shallow( 
    <TodoListItem
    todo = { demoTodos[0] }
    index={ 1 }
    handleToggleTodo= { handleToggleTodo }
    handleDeleteTodo = { handleDeleteTodo }
    /> );

    test('shoul show the component <TodoListItem/> successfully', () => {
        
        expect( wrapper ).toMatchSnapshot();
    });

    test('should call the function handleDelete', () => {
        const button = wrapper.find('button');
        button.simulate('click');
        expect( handleDeleteTodo ).toHaveBeenCalledWith( demoTodos[0] );
    });
    
    test('should call the function handleToggle', () => {
        const paragraph = wrapper.find('p');
        paragraph.simulate('click');
        expect( handleToggleTodo ).toHaveBeenCalledTimes(1);
    });


    test('should show the text successfully', () => {
        const paragraph = wrapper.find('p');
        expect( paragraph.text().trim() ).toBe( `${1}. ${demoTodos[0].desc}`);
    });
    
    
});
