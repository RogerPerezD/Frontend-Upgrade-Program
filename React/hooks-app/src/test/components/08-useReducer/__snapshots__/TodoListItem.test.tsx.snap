// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <TodoListItem/> shoul show the component <TodoListItem/> successfully 1`] = `
<li
  className="list-group-item"
>
  <p
    className="text-center"
    onClick={[Function]}
  >
    1
    . 
    Learn react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
